servus/serializable.cpp

```diff
@@ -23,37 +23,85 @@
 
 namespace servus
 {
+
+class Serializable::Impl
+{
+public:
+    void notifyDeserialized() const
+    {
+        if( deserialized )
+            deserialized();
+    }
+
+    void notifySerialize() const
+    {
+        if( serialize )
+            serialize();
+    }
+
+    Serializable::DeserializeFunc deserialized;
+    Serializable::SerializeFunc serialize;
+};
+
+Serializable::Serializable()
+    : _impl( new Serializable::Impl( ))
+{}
+
+Serializable::~Serializable()
+{
+    delete _impl;
+}
+
 uint128_t Serializable::getTypeIdentifier() const
 {
     return make_uint128( getTypeName( ));
 }
 
-Serializable::ChangeFunc Serializable::setUpdatedFunction(
-    const ChangeFunc& func )
+bool Serializable::fromBinary( const Data& data )
+{
+    return fromBinary( data.ptr.get(), data.size );
+}
+
+bool Serializable::fromBinary( const void* data, const size_t size )
+{
+    if( _fromBinary( data, size ))
+    {
+        _impl->notifyDeserialized();
+        return true;
+    }
+    return false;
+}
+
+Serializable::Data Serializable::toBinary() const
+{
+    _impl->notifySerialize();
+    return _toBinary();
+}
+
+bool Serializable::fromJSON( const std::string& json )
 {
-    const ChangeFunc old = _updated;
-    _updated = func;
-    return old;
+    if( _fromJSON( json ))
+    {
+        _impl->notifyDeserialized();
+        return true;
+    }
+    return false;
 }
 
-Serializable::ChangeFunc Serializable::setRequestedFunction(
-    const ChangeFunc& func )
+std::string Serializable::toJSON() const
 {
-    const ChangeFunc old = _requested;
-    _requested = func;
-    return old;
+    _impl->notifySerialize();
+    return _toJSON();
 }
 
-void Serializable::notifyUpdated() const
+void Serializable::registerDeserializedCallback( const DeserializeFunc& func )
 {
-    if( _updated )
-        _updated();
+    _impl->deserialized = func;
```

Comment:
I forgot about that, and I'm actually not sure about it. That means that if users want to replace the registered function, they can't unless they deregister it first (which at the moment is only possible by registering a void function). We don't have the use case though, so I can implement the throw until multiple callbacks are supported. I'll update the change including also the bool() type callbacks.